ui/ducks/ens.js

```diff
@@ -0,0 +1,179 @@
+
+// Local Constants
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
+const ZERO_X_ERROR_ADDRESS = '0x';
+
+const initialState = {
+  stage: 'UNINITIALIZED',
+  resolution: null,
+  error: null,
+  warning: null,
+  network: null,
+};
+
+export const ensInitialState = initialState;
+
+const name = 'ENS';
+
+let ens = null;
+
+const slice = createSlice({
+  name,
+  initialState,
+  reducers: {
+    processEnsError: (state, action) => {
+      const { recipient, network, reason } = action.payload;
+      if (
+        isValidDomainName(recipient) &&
+        reason.message === 'ENS name not defined.'
+      ) {
+        state.error =
+          network === MAINNET_NETWORK_ID
+            ? ENS_NO_ADDRESS_FOR_NAME
+            : ENS_NOT_FOUND_ON_NETWORK;
+      } else if (reason.message === 'Illegal Character for ENS.') {
+        state.error = ENS_ILLEGAL_CHARACTER;
+      } else {
+        log.error(reason);
+        state.error = ENS_UNKNOWN_ERROR;
+      }
+    },
+    processEnsResult: (state, action) => {
+      state.resolution = action.payload;
+      if (state.resolution === ZERO_ADDRESS) {
+        state.error = ENS_NO_ADDRESS_FOR_NAME;
+      }
+      if (state.resolution === ZERO_X_ERROR_ADDRESS) {
+        state.error = ENS_REGISTRATION_ERROR;
+      }
+      if (
+        isValidDomainName(state.resolution) &&
+        isConfusing(state.resolution)
+      ) {
+        state.warning = CONFUSING_ENS_ERROR;
+      }
+    },
+    enableEnsLookup: (state, action) => {
+      state.stage = 'INITIALIZED';
+      state.error = null;
+      state.resolution = null;
+      state.warning = null;
+      state.network = action.payload;
+    },
+    disableEnsLookup: (state) => {
+      state.stage = 'NO_NETWORK_SUPPORT';
+      state.error = ENS_NOT_SUPPORTED_ON_NETWORK;
+      state.warning = null;
+      state.resolution = null;
+      state.network = null;
+    },
+    resetResolution: (state) => {
+      state.resolution = null;
+      state.warning = null;
+      state.error =
+        state.stage === 'NO_NETWORK_SUPPORT'
+          ? ENS_NOT_SUPPORTED_ON_NETWORK
+          : null;
+    },
+  },
+  extraReducers: (builder) => {
+    builder.addCase(CHAIN_CHANGED, (state, action) => {
+      if (action.payload !== state.currentChainId) {
+        state.stage = 'UNINITIALIZED';
+      }
+    });
+  },
+});
+
+const { reducer, actions } = slice;
+export default reducer;
+
+const {
+  disableEnsLookup,
+  processEnsError,
+  processEnsResult,
+  enableEnsLookup,
+  resetResolution,
+} = actions;
+export { resetResolution };
+
+export function initializeEnsSlice() {
+  return async (dispatch, getState) => {
+    const state = getState();
+    const chainId = getCurrentChainId(state);
+    const network = CHAIN_ID_TO_NETWORK_ID_MAP[chainId];
+    const networkIsSupported = Boolean(networkMap[network]);
+    if (networkIsSupported) {
+      ens = new ENS({ provider: global.ethereumProvider, network });
```

Comment:
We have an ENS controller that does ENS lookups as well. Would probably be better to use that instead.

Edit: mentioned this on the call - this might be left for a later iteration.